### PR TITLE
ssl: SNI server support

### PR DESCRIPTION
**Pull request by fqj1994**
Hello,

This PR adds TLS SNI extension server support to ssl module.

### CODE REVIEW

File: lib/ssl/src/tls_connection.erl [hunk 1]
```diff
@@ -431,7 +440,37 @@ next_state(Current, Next, #ssl_tls{type = ?HANDSHAKE, fragment = Data},
...
+									   diffie_hellman_params = DHParams,
+									   ssl_options = NewOptions,
+									   sni_hostname = Hostname
+									  }
+							 end
+					 end;
+				 _ ->
+					 State0
+			 end,
+	State = State1#state{protocol_buffers =
```
Comment (nox) [inline lib/ssl/src/tls_connection.erl]: I think that piece of code is really badly indented.

**Comment by nox** [comment]
There are no tests, too.

**Comment by IngelaAndin** [comment]
I think the following could be formulated clearer:

"If clients send a client hello with SNI extension, and the hostname
the client gives matches one of your list. The corresponding SSL options
will be merged with the new SSL options in your sni_hosts options."

Maybe something along the lines:

"If the server receives a SNI (Server Name Indication) from the client
matching a host listed in the sni_hosts option, the specified options
for that host will override previously specified options."

I think function sni_hostname could be replaced by making
the sni_hostname part of ssl:connection_info/1 return value
on the server side.

Also we need real test cases not only prof of concept.
You can look at ssl_to_openssl_SUITE.erl in the test directory
and should be able to turn your prof of concept tests
into real tests.

**Comment by fqj1994** [comment]
doc updated about the description and rebased into one commit.
I'll merge ssl:sni_hostname into connection_info/1 this weekend and update corresponding doc.
Tests will be added later not before next week as I have some school work to do.

**Comment by fqj1994** [comment]
But, merging sni_hostname into connection_info/1 (to return `{ok, {Version, CipherSuite, SNIHostname}}`) may break existing application if someone writes something like
```
case ssl:connection_info(S) of
    {ok, {Ver, _}} ->
        Ver;
    {error, R} ->
        throw R
end.
```
I don't know whether it's proper or not to do that.

**Comment by IngelaAndin** [comment]
Hum ... I did not remember the connection_info function looked like that, I just wrongly assumed it was  like other info functions, that was unfortunate.  We could however make a connection_information/[1,2] which works like erlang:process_info and deprecate connection_info/1
as it does not have a flexible design which I think it should have, even though I might be my fault it looks as it does.

**Comment by fqj1994** [comment]
So, should I wait for the new `connection_info` (and then add sni_hosts into it) or implement a new `connection_info` (which also deals with version and ciphersuite) on my own ?

**Comment by IngelaAndin** [comment]
It is not such a big job to implement the new function, but we have a lot of other high prio things to do, so if you do it, it will happen quicker. You do not need to worry about deprecation we can fix that later.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by IngelaAndin** [comment]
I am sorry  if I was unclear but I do not think we want to add a function to ssl_connection for each
information attribute that we want to get about a connection. Also we need to take care of the case that
there is no connection as connection_info does.  I think you should send the list of connection attributes that you want the value for to the connection process that will send you a list back.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by matrixise** [comment]
Ping, are there some news about this PR ?

**Comment by IngelaAndin** [comment]
I have not heard anything since my last comment. We will not take the patch as is. So someone needs to work on it. If nobody thinks its important enough we will take it up later when we find the time or it gets up prioritized  for some reason.

**Comment by fqj1994** [comment]
@IngelaAndin 

Do you mean anything like commit  413da16 for connection_information ?

Thx

**Comment by IngelaAndin** [comment]
Yes that was what I meant, that was strange it seems to be there now but I can not recall getting information about it being added. I do apologies for missing it.  So the ball is ours again.

**Comment by IngelaAndin** [comment]
Looks ok,  but we want this function to be documented instead of sni_hostname/1. It would also
be nice if you could implement the current connection_info with help of this new function to reduce
code complexity.   

Deprecation of connection_info function goes in the file otp/lib/stdlib/src/otp_internal.erl if you have any trouble with that I can as I said add that on top of your commits. When we have a deprecated warning we can remove the documentation of the old function.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by fqj1994** [comment]
Hello,

docs updated.

Does the deprecation of functions look like this: (https://github.com/fqj1994/otp/commit/abd1fea883cc2f1dd28a7606f3aeb821586fee93#diff-fa2ff57a5d26d8bfb0f955698930c7c0R583)

**Comment by OTP-Maintainer** [comment]
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: ssl: add docs about SNI server support, and connection_information


----------------

I am a script, I am not human 

----------------

**Comment by IngelaAndin** [comment]
Yes deprecation looks correct.

**Comment by matrixise** [comment]
really happy to see the progress of this bug, I will use Erlang with the SNI support for a project.
Thank you a lot

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by fqj1994** [comment]
@IngelaAndin 

I'll prepare for the proposal presentation of the thesis for the bachelor's degree for the next 10 days. Remaining parts (tests) will be committed later.

And, I'm planing to move the sni_hosts (https://github.com/fqj1994/otp/commit/a86ce3369f0e9b0b316d04d032d059b109ec7360#diff-79b520311347e34fac9cffbc4ea373f9R673) atom from L673 to L671(after next_protocol_advertised) for auto-merge. This is the only conflict parts aginst the current master branch. Is that ok? If yes, should this commit be a standalone commit or squash it into the previous commit?

Thanks

**Comment by essen** [comment]
Please tell me if I can help, I need this in OTP 18.

**Comment by OTP-Maintainer** [comment]
./otp_build all -a failed:

    ce/otp/lib/public_key/doc/src'
    make[3]: Nothing to be done for `opt'.
    make[3]: Leaving directory `/ldisk/jenkins/jobs/pullrequests/workspace/otp/lib/public_key/doc/src'
    === Leaving application public_key
    make[2]: Leaving directory `/ldisk/jenkins/jobs/pullrequests/workspace/otp/lib/public_key'
    make[2]: Entering directory `/ldisk/jenkins/jobs/pullrequests/workspace/otp/lib/ssl'
    === Entering application ssl
    make[3]: Entering directory `/ldisk/jenkins/jobs/pullrequests/workspace/otp/lib/ssl/src'
     ERLC	../ebin/ssl_session_cache_api.beam
     ERLC	../ebin/ssl.beam
    ssl.erl:299: spec for ssl:connection_information/2 already defined
    make[3]: *** [../ebin/ssl.beam] Error 1
    make[3]: Leaving directory `/ldisk/jenkins/jobs/pullrequests/workspace/otp/lib/ssl/src'
    make[2]: *** [opt] Error 2
    make[2]: Leaving directory `/ldisk/jenkins/jobs/pullrequests/workspace/otp/lib/ssl'
    make[1]: *** [opt] Error 2
    make[1]: Leaving directory `/ldisk/jenkins/jobs/pullrequests/workspace/otp/lib'
    make: *** [libs] Error 2



----------------

I am a script, I am not human 

----------------

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by IngelaAndin** [comment]
I think that relocation of atom for automatic merge may be squashed, it adds no own value.
In the documentation I think you need to add the extended explanation to the description (desc) as well as the  function summery (fsummary).  Also I did merge new things to master today (it should be visible soon if is not already) so you will need to rebase.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by fqj1994** [comment]
1st part of tests testing 3 different circumstances (no sni header, sni match, sni no match found) using SSL server and client in erlang/otp.
The 2nd part of interacting with openssl will be added later next week (after the presentation)

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by fqj1994** [comment]
@IngelaAndin 
all tests I can think of has been finished.

**Comment by IngelaAndin** [comment]
We are getting close to merge now, but there are some follow ups of test cases and integration left to do.

**Comment by fqj1994** [comment]
@IngelaAndin 
Anything I can help ?

**Comment by IngelaAndin** [comment]
Hoops ... I commented the wrong PR (many ssl activites at once) sorry, but the statement is not entirely incorrect.  And we are looking into a couple of tests case that  fail sometimes on some  platforms. I have merged my own CRL work to master and the ALPN PR request will be merged today. 
So this one will be the next one that we will want to merge.  One thing you can do is to rebase your branch onto the new master tomorrow, and we will take it from there.

**Comment by fqj1994** [comment]
It looks the PR is against `maint` branch. Should I close this PR and file another one against `master` branch ?

**Comment by IngelaAndin** [comment]
Well yes we would like it for master. It is too little time to get it in to the next maint release and as it is a
new feature it is preferable to put it into the next major release.

**Comment by matrixise** [comment]
when you say "major", in 18.x ? and not 17.5 ?

**Comment by IngelaAndin** [comment]
Yes 18 is the next major release.

**Comment by matrixise** [comment]
do you have a roadmap for R18?

**Comment by IngelaAndin** [comment]
OTP 18.0 is planned for the end of  June.

**Comment by matrixise** [comment]
thanks too late for me, but really thank you for this incredible job.

**Comment by IngelaAndin** [comment]
I suppose you can always have your own patch for the 17 track that you can drop once you can upgrade to 18.

**Comment by matrixise** [comment]
yep, of course, don't worry about that, and you are right, it's a new feature and not a bugfix.

**Comment by fqj1994** [comment]
Reopened at https://github.com/erlang/otp/pull/656 for master branch.

**Pull request closed without merging**